File: benchmark/python/bench.py
import json
import os
import sys
import time

import zen

root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
fixtures = os.path.join(root, "fixtures")
manifest = json.load(open(os.path.join(root, "manifest.json")))
iters = int(os.environ.get("BENCH_ITERS", "2000"))
out_path = sys.argv[1] if len(sys.argv) > 1 else None


def loader(key):
    with open(os.path.join(fixtures, key)) as f:
        return f.read()


engine = zen.ZenEngine({"loader": loader})

results = []
for e in manifest:
    for trace in (False, True):
        opts = {"trace": trace}
        engine.evaluate(e["file"], e["input"], opts)
        start = time.perf_counter_ns()
        for _ in range(iters):
            engine.evaluate(e["file"], e["input"], opts)
        per = (time.perf_counter_ns() - start) / iters
        suffix = " +trace" if trace else ""
        results.append({"name": f'{e["name"]} ({e["kind"]}){suffix}', "unit": "ns/op", "value": per})

if out_path:
    with open(out_path, "w") as f:
        f.write(json.dumps(results, indent=2))
else:
    for r in results:
        print(f'{r["name"]:<44} {r["value"]:>12.0f} ns/op')

File: benchmark/node/bench.mjs
import { readFile } from 'node:fs/promises';
import { readFileSync, writeFileSync } from 'node:fs';
import { hrtime, argv, env } from 'node:process';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { createRequire } from 'node:module';

const require = createRequire(import.meta.url);
const { ZenEngine } = require('../../bindings/nodejs/index.js');

const dir = path.dirname(fileURLToPath(import.meta.url));
const root = path.join(dir, '..');
const fixtures = path.join(root, 'fixtures');
const manifest = JSON.parse(readFileSync(path.join(root, 'manifest.json'), 'utf8'));
const iters = Number(env.BENCH_ITERS || 2000);
const outPath = argv[2];

const loader = (key) => readFile(path.join(fixtures, key));
const engine = new ZenEngine({ loader });

const results = [];
for (const e of manifest) {
  await engine.evaluate(e.file, e.input);
  const start = hrtime.bigint();
  for (let i = 0; i < iters; i++) {
    await engine.evaluate(e.file, e.input);
  }
  const per = Number(hrtime.bigint() - start) / iters;
  results.push({ name: `${e.name} (${e.kind})`, unit: 'ns/op', value: per });
}

engine.dispose?.();

const json = JSON.stringify(results, null, 2);
if (outPath) writeFileSync(outPath, json);
else console.log(json);
